tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php

```diff
@@ -0,0 +1,348 @@
+		'tests20/line0497'           => "Closing P tag implicitly creates opener, which we don't visit.",
+		'tests23/line0001'           => 'Unimplemented: Reconstruction of active formatting elements.',
+		'tests23/line0041'           => 'Unimplemented: Reconstruction of active formatting elements.',
+		'tests23/line0069'           => 'Unimplemented: Reconstruction of active formatting elements.',
+		'tests23/line0101'           => 'Unimplemented: Reconstruction of active formatting elements.',
+		'tests25/line0169'           => 'Bug.',
+		'tests26/line0263'           => 'Bug: An active formatting element should be created for a trailing text node.',
+		'tests8/line0001'            => 'Bug.',
+		'tests8/line0020'            => 'Bug.',
+		'tests8/line0037'            => 'Bug.',
+		'tests8/line0052'            => 'Bug.',
+		'webkit01/line0174'          => 'Bug.',
+	);
+
+
+	/**
+	 * Verify the parsing results of the HTML Processor against the
+	 * test cases in the Html5lib tests project.
+	 *
+	 * @dataProvider data_external_html5lib_tests
+	 *
+	 * @param string $fragment_context Context element in which to parse HTML, such as BODY or SVG.
+	 * @param string $html             Given test HTML.
+	 * @param string $result           Tree structure of parsed HTML.
+	 */
+	public function test_parse( $fragment_context, $html, $result ) {
+		if ( self::SKIP_HEAD_TESTS ) {
+			$html_start = "<html>\n  <head>\n  <body>\n";
+			if (
+				strlen( $result ) < strlen( $html_start ) ||
+				substr( $result, 0, strlen( $html_start ) ) !== $html_start
+			) {
+				$this->markTestSkipped( 'Skip test with expected content in <head> (unsupported).' );
+			}
+		}
+
+		if ( array_key_exists( $this->dataName(), self::SKIP_TESTS ) ) {
+			$this->markTestSkipped( self::SKIP_TESTS[ $this->dataName() ] );
+		}
+
+		$processed_tree = self::build_tree_representation( $fragment_context, $html );
+
+		if ( null === $processed_tree ) {
+			$this->markTestIncomplete( 'Test includes unsupported markup.' );
+		}
+
+		$this->assertEquals( $result, $processed_tree, "HTML was not processed correctly:\n{$html}" );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * Tests from https://github.com/html5lib/html5lib-tests
+	 *
+	 * @return array[]
+	 */
+	public function data_external_html5lib_tests() {
+		$test_dir = DIR_TESTDATA . '/html5lib-tests/tree-construction/';
+
+		$handle = opendir( $test_dir );
+		while ( false !== ( $entry = readdir( $handle ) ) ) {
+			if ( ! stripos( $entry, '.dat' ) ) {
+				continue;
+			}
+
+			if ( 'entities01.dat' === $entry || 'entities02.dat' === $entry ) {
+				continue;
+			}
+
+			foreach ( self::parse_html5_dat_testfile( $test_dir . $entry ) as $k => $test ) {
+				// strip .dat extension from filename
+				$test_suite = substr( $entry, 0, -4 );
+				$line       = str_pad( strval( $test[0] ), 4, '0', STR_PAD_LEFT );
+
+				yield "{$test_suite}/line{$line}" => array_slice( $test, 1 );
```

Comment:
We could abstract the two `markTestSkipped()` blocks from the test method into a `should_skip_test()` method, such as:

```php
private static function should_skip_test( $name, $result ) {
	if ( array_key_exists( $name, self::SKIP_TESTS ) ) {
		return true;
	}

	if ( ! self::SKIP_HEAD_TESTS ) {
		return false;
	}

	$html_start        = "<html>\n  <head>\n  <body>\n";
	$html_start_length = strlen( $html_start );
	return strlen( $result ) < $html_start_length || substr( $result, 0, $html_start_length ) ) !== $html_start;
}
```

Then we could change the `yield` to:
```php
$dataset_name = "{$test_suite}/line{$line}";
$test_data    = array_slice( $test, 1 );

if ( ! self::should_skip_test( $dataset_name, $test_data[2] ) ) {
	yield $dataset_name => $test_data;
}
```

This would:
- reduce bloat in the data provider
- reduce "unnecessary" output in the test results for skipped tests
- isolate the logic for skipping them

While it wouldn't cover the `markTestIncomplete()` datasets, it could at least cover the 1109 skipped ones.